File: bindings/java/src/test/java/ai/stogas/verifier/TransportTest.java
package ai.stogas.verifier;

import java.util.Map;
import java.util.concurrent.CompletableFuture;
import org.junit.jupiter.api.Test;
import static org.junit.jupiter.api.Assertions.*;
import static org.junit.jupiter.api.Assumptions.*;

class TransportTest {
    @Test void rejectsInvalidConfigurationWithNativeCodes() {
        for (var options : java.util.List.of(Map.of("security", "unsupported"), Map.of("environment", "unsupported"), Map.of("max_connections", 0))) {
            var error = assertThrows(Transport.VerificationException.class, () -> new Transport(options));
            assertFalse(error.code().isEmpty());
        }
    }
    @Test void realTransportClosesOnceAndGuardsCalls() {
        assumeTrue("1".equals(System.getenv("STOGAS_NATIVE_STAGING_TEST")));
        Transport transport;
        try (var opened = new Transport(Map.of("environment", "staging", "security", "e2ee"))) {
            transport = opened;
            assertEquals("127.0.0.1", opened.baseUrl().getHost());
            var one = CompletableFuture.runAsync(opened::refresh);
            var two = CompletableFuture.runAsync(opened::refresh);
            CompletableFuture.allOf(one, two).join();
        }
        transport.close();
        assertThrows(IllegalStateException.class, transport::refresh);
    }
}

File: bindings/ruby/test/transport_test.rb
require 'minitest/autorun'
require 'stogas_verifier'

class TransportTest < Minitest::Test
  def test_invalid_options_preserve_native_error_codes
    [{security: 'unsupported'}, {environment: 'unsupported'}, {max_connections: 0}].each do |options|
      error = assert_raises(Stogas::VerificationError) { Stogas::Transport.new(**options) }
      refute_empty error.code
    end
  end

  def test_native_transport_lifetime
    skip 'requires explicit staging evidence qualification' unless ENV['STOGAS_NATIVE_STAGING_TEST'] == '1'
    transport = nil
    assert_raises(RuntimeError) do
      Stogas::Transport.open(environment: 'staging', security: 'e2ee') do |opened|
        transport = opened
        assert_match(%r{\Ahttp://127\.0\.0\.1:\d+/}, transport.base_url)
        threads = 3.times.map { Thread.new { transport.refresh } }
        threads.each(&:value)
        raise 'application failed'
      end
    end
    transport.close
    assert_raises(IOError) { transport.refresh }
    GC.start
  end
end
